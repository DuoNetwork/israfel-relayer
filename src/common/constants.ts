export const DB_ISRAFEL = 'israfel';
export const DB_LIVE_ORDERS = 'liveOrders';
export const DB_RAW_ORDERS = 'rawOrders';
export const DB_USER_ORDERS = 'userOrders';
export const DB_SEQUENCE = 'sequence';
export const DB_STATUS = 'status';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PAIR = 'pair';
export const DB_INITIAL_SEQ = 'initialSequence';
export const DB_CURRENT_SEQ = 'currentSequence';
export const DB_ORDER_HASH = 'orderHash';
export const DB_PRICE = 'price';
export const DB_ACCOUNT = 'account';
export const DB_ACCOUNT_YM = 'accountYearMonth';
export const DB_PAIR_OH_SEQ_STATUS = 'pairOrderHashSequenceStatus';
export const DB_TYPE = 'type';
export const DB_CONFIRMED = 'confirmed';
export const DB_MATCHING = 'matching';
export const DB_USER = 'user';
export const DB_ORDER_PROCESSOR = 'orderProcessor';
export const DB_ORDER_MATCHER = 'orderMatcher';
export const DB_ORDER_WATCHER = 'orderWatcher';
export const DB_RELAYER = 'relayer';
export const DB_URL = 'url';
export const DB_SIDE = 'side';
export const DB_ASK = 'ask';
export const DB_BID = 'bid';
export const DB_BALANCE = 'balance';
export const DB_AMOUNT = 'amount';
export const DB_FILL = 'fill';
export const DB_PFILL = 'pFill';
export const DB_CREATED_AT = 'createdAt';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_UPDATED_BY = 'updatedBy';
export const DB_PROCESSED = 'processed';
export const DB_TOKENS = 'tokens';
export const DB_ADDRESS = 'address';
export const DB_CODE = 'code';
export const DB_EXP = 'expiry';
export const DB_DENOMINATION = 'denomination';
export const DB_PRECISION = 'precision';
export const DB_FEE = 'fee';
export const DB_FEE_ASSET = 'feeAsset';

export const DB_ADD = 'add';
export const DB_UPDATE = 'update';
export const DB_TERMINATE = 'terminate';

export const DB_PROCESS = 'process';
export const DB_HOSTNAME = 'hostname';
export const DB_COUNT = 'count';

export const DB_0X_SENDER_ADDR = 'senderAddress';
export const DB_0X_MAKER_FEE = 'makerFee';
export const DB_0X_TAKER_FEE = 'takerFee';
export const DB_0X_MAKER_ADDR = 'makerAddress';
export const DB_0X_TAKER_ADDR = 'takerAddress';
export const DB_0X_MAKER_ASSET_AMT = 'makerAssetAmount';
export const DB_0X_TAKER_ASSET_AMT = 'takerAssetAmount';
export const DB_0X_MAKER_ASSET_DATA = 'makerAssetData';
export const DB_0X_TAKER_ASSET_DATA = 'takerAssetData';
export const DB_0X_SALT = 'salt';
export const DB_0X_EXCHANGE_ADDR = 'exchangeAddress';
export const DB_0X_FEE_RECIPIENT_ADDR = 'feeRecipientAddress';
export const DB_0X_EXPIRATION_TIME_SECONDS = 'expirationTimeSeconds';
export const DB_0X_SIGNATURE = 'signature';

export const DB_ORDERS = 'orders';
export const DB_ORDER_BOOKS = 'orderBooks';
export const DB_QUEUE = 'queue';
export const DB_CACHE = 'cache';
export const DB_PUBSUB = 'pubsub';
export const DB_SNAPSHOT = 'snapshot';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';

export const TAKER_ETH_DEPOSIT = 10; // for development only
export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const RELAYER_ADDR_KOVAN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const RELAYER_ADDR_MAIN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const DUMMY_ADDR = '0x0000000000000000000000000000000000000000';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const NETWORK_ID_MAIN = 1;
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const PENDING_HOURS = 24;

export const WS_HISTORY = 'history';
export const WS_INFO = 'info';
export const WS_SUB = 'subscribe';
export const WS_UNSUB = 'unsubscribe';
export const WS_CHANNEL = 'channel';
export const WS_METHOD = 'method';
export const WS_INVALID_REQ = 'invalid request';
export const WS_INVALID_ORDER = 'invalid order';
export const WS_ERROR = 'error';
export const WS_OK = 'ok';
export const WS_SERVICE_NA = 'service not available';

export const WS_CHANNEL_ORDERBOOK = 'orderbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const TOKEN_WETH = 'WETH';

export const PRUNE_INTERVAL = 3600000;
export const ORDER_PRUNE = 'pruneOrder';
export const SET_ALLOWANCE = 'setAllowance';
export const START_RELAYER = 'startRelayer';

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const PRICE_PRECISION = 8;

export const BASE_DERIVATION_PATH = `44'/60'/0'/0`;
