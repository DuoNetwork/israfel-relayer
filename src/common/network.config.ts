
export const mnemonicConfig = {
	MNEMONIC: 'alice bob charles david eric frank george henry is jack kobe let',
	BASE_DERIVATION_PATH: `44'/60'/0'/0`
}

export const KOVAN_CONFIGS = {
	rpcUrl: 'https://kovan.infura.io/WSDscoNUvMiL1M7TvMNP',
	networkId: 42,
};
