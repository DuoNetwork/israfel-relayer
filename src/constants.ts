export const DB_ORDERS = 'orders';
export const DB_MAKER_ADDR = 'makerAddr';
export const DB_TIMESTAMP = 'timestamp';
export const DB_UPDATED_AT = 'updatedAt';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const RELAYER_HTTP_URL = 'http://localhost:3000/v0';
export const RELAYER_WS_URL = 'ws://localhost:8080';
export const MNEMONIC = 'concert load couple harbor equip island argue ramp clarify fence smart topic'
export const BASE_DERIVATION_PATH =  `44'/60'/0'/0`;
export const PENDING_HOURS = 24;

export const WS_CHANNEL_ORDERBOOK = 'orerbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
