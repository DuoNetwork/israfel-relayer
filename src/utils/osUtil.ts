import * as os from 'os';

class OsUtil {
	public getHostName() {
		return os.hostname();
	}
}

const osUtil = new OsUtil();
export default osUtil;
