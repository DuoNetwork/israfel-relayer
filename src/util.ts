import { BigNumber } from '0x.js';
import moment from 'moment';
import * as os from 'os';
import * as CST from './constants';
import { IOption } from './types';

class Util {
	public logLevel: string = CST.LOG_INFO;

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	private log(text: any, level: string): void {
		if (CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public getHostName() {
		return os.hostname();
	}

	public getRandomFutureDateInSeconds() {
		return new BigNumber(Date.now() + CST.TEN_MINUTES_MS).div(CST.ONE_SECOND_MS).ceil();
	}

	public stringToBN(value: string): BigNumber {
		return new BigNumber(value);
	}

	public defaultOption: IOption = {
		live: false,
		token: 'ZRX',
		amount: 1,
		maker: 0,
		spender: 1
	};

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public parseOptions(argv: string[]): IOption {
		const option: IOption = this.defaultOption;

		for (let i = 3; i < argv.length; i++) {
			option.live = process.argv.includes('live');
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'token':
					option.token = args[1] || option.token;
					break;
				case 'amount':
					option.amount = Number(args[1]) || option.amount;
					break;
				case 'maker':
					option.maker = Number(args[1]) || option.maker;
					break;
				case 'spender':
					option.maker = Number(args[1]) || option.spender;
					break;
				default:
					break;
			}
		}

		return option;
	}
}

const util = new Util();
export default util;
