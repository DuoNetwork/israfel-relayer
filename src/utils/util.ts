import { Constants } from '@finbook/israfel-common';
import { IOption } from '../common/types';

export default class Util {
	public static defaultOption: IOption = {
		env: Constants.DB_DEV,
		tokens: [],
		token: '',
		debug: false,
		server: false
	};

	public static parseOptions(argv: string[]): IOption {
		const option: IOption = this.defaultOption;
		option.server = argv.includes('server');
		option.debug = argv.includes('debug');
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'env':
					option.env = [Constants.DB_LIVE, Constants.DB_UAT].includes(args[1])
						? args[1]
						: option.env;
					break;
				case 'tokens':
					option.tokens = args[1].split(',');
					break;
				case 'token':
					option.token = args[1] || option.token;
					break;
				default:
					break;
			}
		}

		return option;
	}
}
