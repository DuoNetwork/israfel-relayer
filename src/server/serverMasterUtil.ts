import child_process from 'child_process';
import { IOption, ISubProcess, IToken } from '../common/types';
import dynamoUtil from '../utils/dynamoUtil';
import util from '../utils/util';
import Web3Util from '../utils/Web3Util';

class ServerMasterUtil {
	public subProcesses: { [key: string]: ISubProcess } = {};

	public async startLaunching(
		web3Util: Web3Util,
		tool: string,
		option: IOption,
		startServer: (web3Util: Web3Util, option: IOption) => any
	): Promise<void> {
		const tokens: IToken[] = await dynamoUtil.scanTokens();

		if (!option.token) {
			util.logInfo('launching all pairs ' + tokens.map(token => token.code).join(','));
			for (const token of tokens) {
				await util.sleep(1000);
				this.subProcesses[token.code] = {
					token: token.code,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchTokenPair(tool, token.code, option);
			}
		} else {
			const token = tokens.find(t => t.code === option.token);
			if (!token) throw new Error('invalid token specified');

			util.logInfo(
				`[${option.token}]:` + 'start launching orderBookServer for pair ' + option.token
			);
			startServer(web3Util, option);
		}
	}

	public launchTokenPair(tool: string, token: string, option: IOption) {
		const cmd =
			`npm run ${tool} token=${token} ${option.server ? ' server' : ''}${
				option.debug ? ' debug' : ''
			}` +
			(process.platform === 'win32' ? ' >>' : ' &>') +
			` ${tool}.${token}.log`;

		util.logInfo(`[${token}]: ${cmd}`);

		const procInstance = child_process.exec(
			cmd,
			process.platform === 'win32' ? {} : { shell: '/bin/bash' }
		);

		this.subProcesses[token].instance = procInstance;
		this.subProcesses[token].lastFailTimestamp = util.getUTCNowTimestamp();

		if (!procInstance) {
			util.logError('Failed to launch ' + token);
			this.retry(tool, option, token);
		} else {
			util.logInfo(`[${token}]: Launched ${tool}`);
			procInstance.on('exit', code => {
				util.logError(`[${token}]: Exit with code ${token}`);
				if (code) this.retry(tool, option, token);
			});
		}
	}

	public retry(tool: string, option: IOption, token: string) {
		const now: number = util.getUTCNowTimestamp();

		if (now - this.subProcesses[token].lastFailTimestamp < 30000)
			this.subProcesses[token].failCount++;
		else this.subProcesses[token].failCount = 1;

		this.subProcesses[token].lastFailTimestamp = now;

		if (this.subProcesses[token].failCount < 3)
			setTimeout(() => this.launchTokenPair(tool, token, option), 5000);
		else util.logError('Retry Aborted ' + token);
	}
}

const serverMasterUtil = new ServerMasterUtil();
export default serverMasterUtil;
