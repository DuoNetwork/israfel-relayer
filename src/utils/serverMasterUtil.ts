import child_process from 'child_process';
import { IOption, ISubProcess, IToken } from '../common/types';
import dynamoUtil from './dynamoUtil';
import osUtil from './osUtil';
import util from './util';

class ServerMasterUtil {
	public subProcesses: { [key: string]: ISubProcess } = {};

	public async startLaunching(
		tool: string,
		option: IOption,
		startServer?: (option: IOption) => any
	): Promise<void> {
		const tokens: IToken[] = await dynamoUtil.scanTokens();

		if (option.token && startServer) {
			const rawToken = tokens.find(t => t.code === option.token);
			if (!rawToken) throw new Error('invalid token specified');

			util.logInfo(`[${option.token}]: start launching ${tool}`);
			startServer(option);
		} else if (!option.tokens.length) {
			util.logInfo('launching all pairs ' + tokens.map(token => token.code).join(','));
			for (const token of tokens) {
				util.logInfo(`[${token.code}]: start launching ${tool}`);
				await util.sleep(1000);
				this.subProcesses[token.code] = {
					token: token.code,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchTokenPair(tool, token.code, option);
			}
		} else if (option.tokens.length)
			for (const token of tokens) {
				if (!option.tokens.includes(token.code)) continue;

				util.logInfo(`[${token.code}]: start launching ${tool}`);
				await util.sleep(1000);
				this.subProcesses[token.code] = {
					token: token.code,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchTokenPair(tool, token.code, option);
			}
	}

	public launchTokenPair(tool: string, token: string, option: IOption) {
		const cmd =
			`npm run ${tool} token=${token} env=${option.env} ${option.server ? ' server' : ''}${
				option.debug ? ' debug' : ''
			}` +
			(osUtil.isWindows() ? ' >>' : ' &>') +
			` ${tool}.${token}.log`;

		util.logInfo(`[${token}]: ${cmd}`);

		const procInstance = child_process.exec(
			cmd,
			osUtil.isWindows() ? {} : { shell: '/bin/bash' }
		);

		this.subProcesses[token].instance = procInstance;
		this.subProcesses[token].lastFailTimestamp = util.getUTCNowTimestamp();

		if (!procInstance) {
			util.logError('Failed to launch ' + token);
			this.retry(tool, option, token);
		} else {
			util.logInfo(`[${token}]: Launched ${tool}`);
			procInstance.on('exit', code => {
				util.logError(`[${token}]: Exit with code ${token}`);
				if (code) this.retry(tool, option, token);
			});
		}
	}

	public retry(tool: string, option: IOption, token: string) {
		const now: number = util.getUTCNowTimestamp();

		if (now - this.subProcesses[token].lastFailTimestamp < 30000)
			this.subProcesses[token].failCount++;
		else this.subProcesses[token].failCount = 1;

		this.subProcesses[token].lastFailTimestamp = now;

		if (this.subProcesses[token].failCount < 3)
			setTimeout(() => this.launchTokenPair(tool, token, option), 5000);
		else util.logError('Retry Aborted ' + token);
	}
}

const serverMasterUtil = new ServerMasterUtil();
export default serverMasterUtil;
