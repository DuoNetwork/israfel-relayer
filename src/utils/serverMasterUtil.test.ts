import child_process from 'child_process';
import { DB_DEV, DB_LIVE } from '../common/constants';
import osUtil from './osUtil';
import serverMasterUtil from './serverMasterUtil';
import util from './util';

test('retry after long enought time', () => {
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: util.getUTCNowTimestamp() - (30000 + 1),
		failCount: 2,
		instance: undefined as any
	};
	serverMasterUtil.retry(
		'tool',
		{
			server: false,
			debug: false,
			live: false
		} as any,
		'token'
	);
	expect(serverMasterUtil.subProcesses['token']).toMatchSnapshot();
});

test('retry within short time', () => {
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: util.getUTCNowTimestamp() - (30000 - 1),
		failCount: 2,
		instance: undefined as any
	};
	serverMasterUtil.retry(
		'tool',
		{
			env: DB_DEV,
			debug: false,
			live: false
		} as any,
		'token'
	);
	expect(serverMasterUtil.subProcesses['token']).toMatchSnapshot();
});

test('launchTokenPair fail win32', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token', {
		env: DB_DEV,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect((serverMasterUtil.retry as jest.Mock<void>).mock.calls).toMatchSnapshot();
});

test('launchTokenPair success win 32', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn(() => {
		return {
			on: jest.fn()
		};
	}) as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 5,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		env: DB_DEV,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect(
		(serverMasterUtil.subProcesses['token'].instance.on as jest.Mock<void>).mock.calls
	).toMatchSnapshot();
});

test('launchTokenPair debug win32', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		env: DB_DEV,
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
});

test('launchTokenPair fail not win32', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token', {
		env: DB_LIVE,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect((serverMasterUtil.retry as jest.Mock<void>).mock.calls).toMatchSnapshot();
});

test('launchTokenPair success not win32', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn(() => {
		return {
			on: jest.fn()
		};
	}) as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 5,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		env: DB_LIVE,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect(
		(serverMasterUtil.subProcesses['token'].instance.on as jest.Mock<void>).mock.calls).toMatchSnapshot()
});

test('launchTkenPair debug not win32', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		env: DB_LIVE,
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
});
