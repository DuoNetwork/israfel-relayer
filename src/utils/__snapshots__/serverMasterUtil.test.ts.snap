// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`launchTkenPair debug not win32 1`] = `
Array [
  Array [
    "npm run tool token=token live  debug &> tool.token.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchTkenPair debug not win32 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair debug win32 1`] = `
Array [
  Array [
    "npm run tool token=token dev  debug >> tool.token.log",
    Object {},
  ],
]
`;

exports[`launchTokenPair debug win32 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair fail not win32 1`] = `
Array [
  Array [
    "npm run tool token=token live  &> tool.token.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchTokenPair fail not win32 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair fail not win32 3`] = `
Array [
  Array [
    "tool",
    Object {
      "debug": false,
      "env": "live",
      "live": false,
    },
    "token",
  ],
]
`;

exports[`launchTokenPair fail win32 1`] = `
Array [
  Array [
    "npm run tool token=token dev  >> tool.token.log",
    Object {},
  ],
]
`;

exports[`launchTokenPair fail win32 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair fail win32 3`] = `
Array [
  Array [
    "tool",
    Object {
      "debug": false,
      "env": "dev",
      "live": false,
    },
    "token",
  ],
]
`;

exports[`launchTokenPair success not win32 1`] = `
Array [
  Array [
    "npm run tool token=token live  &> tool.token.log",
    Object {
      "shell": "/bin/bash",
    },
  ],
]
`;

exports[`launchTokenPair success not win32 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair success not win32 3`] = `
Array [
  Array [
    "exit",
    [Function],
  ],
]
`;

exports[`launchTokenPair success win 32 1`] = `
Array [
  Array [
    "npm run tool token=token dev  >> tool.token.log",
    Object {},
  ],
]
`;

exports[`launchTokenPair success win 32 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair success win 32 3`] = `
Array [
  Array [
    "exit",
    [Function],
  ],
]
`;

exports[`retry after long enought time 1`] = `
Object {
  "failCount": 1,
  "instance": undefined,
  "lastFailTimestamp": 1234567890,
  "token": "token",
}
`;

exports[`retry within short time 1`] = `
Object {
  "failCount": 3,
  "instance": undefined,
  "lastFailTimestamp": 1234567890,
  "token": "token",
}
`;
